src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Windows.cs

```diff
@@ -341,29 +339,24 @@ private static PingReply CreatePingReplyFromIcmpEchoReply(in Interop.IpHlpApi.IC
         {
             const int DontFragmentFlag = 2;
 
-            IPAddress address = new IPAddress(reply.address);
-            IPStatus ipStatus = GetStatusFromCode((int)reply.status);
+            var address = new IPAddress(reply.address);
+            var ipStatus = GetStatusFromCode((int)reply.status);
 
-            long rtt;
-            PingOptions? options;
-            byte[] buffer;
+            var (rtt, options, buffer) = ipStatus == IPStatus.Success
+                ? (reply.roundTripTime, 
+                   new PingOptions(reply.options.ttl, (reply.options.flags & DontFragmentFlag) > 0),
+                   CreateBuffer(reply.data, reply.dataSize))
+                : (0L, null, Array.Empty<byte>());
 
-            if (ipStatus == IPStatus.Success)
-            {
-                // Only copy the data if we succeed w/ the ping operation.
-                rtt = reply.roundTripTime;
-                options = new PingOptions(reply.options.ttl, (reply.options.flags & DontFragmentFlag) > 0);
-                buffer = new byte[reply.dataSize];
-                Marshal.Copy(reply.data, buffer, 0, reply.dataSize);
-            }
-            else
+            return new PingReply(address, options, ipStatus, rtt, buffer);
+
+            static byte[] CreateBuffer(IntPtr data, int dataSize)
             {
-                rtt = 0;
-                options = null;
-                buffer = Array.Empty<byte>();
+                Span<byte> bufferSpan = stackalloc byte[dataSize];
+                Marshal.Copy(data, bufferSpan, 0, dataSize);
+    
+                return bufferSpan.ToArray();
             }
```

Comment:
This method is pure regression. It's still allocating a new array, and copying the data twice with a temporary span.